cpp/tests/sparse/preprocess.cu

```diff
@@ -0,0 +1,299 @@
+        handle, ab_rows.view(), ab_columns.view(), ab_values.view(), num_rows * 2, num_cols);
+      sparseEncoder->fit(handle, a_matrix);
+      sparseEncoder->fit(handle, b_matrix);
+      sparseEncoder->save(handle, "test_save.txt");
+      sparseEncoder =
+        raft::sparse::matrix::loadSparseEncoder<Type_f, Index_>(handle, "test_save.txt");
+      sparseEncoder->transform(handle, c_matrix, result.data_handle(), bm25_on);
+    } else {
+      sparseEncoder->fit(handle, csr_matrix);
+      sparseEncoder->transform(handle, csr_matrix, result.data_handle(), bm25_on);
+    }
+    delete sparseEncoder;
+
+    if (bm25_on) {
+      auto bm25_vals = raft::make_device_vector<Type_f, int64_t>(handle, int(ab_values.size()));
+      raft::util::calc_tfidf_bm25<Index_, Type_f>(handle, csr_matrix.view(), bm25_vals.view());
+      ASSERT_TRUE(raft::devArrMatch<Type_f>(bm25_vals.data_handle(),
+                                            result.data_handle(),
+                                            result.size(),
+                                            raft::CompareApprox<Type_f>(2e-5),
+                                            stream));
+    } else {
+      auto tfidf_vals = raft::make_device_vector<Type_f, int64_t>(handle, int(ab_values.size()));
+      raft::util::calc_tfidf_bm25<Index_, Type_f>(
+        handle, csr_matrix.view(), tfidf_vals.view(), true);
+      ASSERT_TRUE(raft::devArrMatch<Type_f>(tfidf_vals.data_handle(),
+                                            result.data_handle(),
+                                            result.size(),
+                                            raft::CompareApprox<Type_f>(2e-5),
+                                            stream));
+    }
+
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  }
+
+ protected:
+  raft::resources handle;
+  cudaStream_t stream;
+
+  SparsePreprocessInputs<Type_f, Index_> params;
+};
+
+using SparsePreprocessFF = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessFF, Result) { Run(false, false); }
+
+using SparsePreprocessTT = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessTT, Result) { Run(true, true); }
+
+using SparsePreprocessFT = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessFT, Result) { Run(false, true); }
+
+using SparsePreprocessTF = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessTF, Result) { Run(true, false); }
+
+using SparsePreprocessBigFF = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessBigFF, Result) { Run(false, false); }
+
+using SparsePreprocessBigTT = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessBigTT, Result) { Run(true, true); }
+
+using SparsePreprocessBigFT = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessBigFT, Result) { Run(false, true); }
+
+using SparsePreprocessBigTF = SparsePreprocessCSR<float, int>;
+TEST_P(SparsePreprocessBigTF, Result) { Run(true, false); }
+
+const std::vector<SparsePreprocessInputs<float, int>> sparse_preprocess_inputs = {
+  {
```

Comment:
Since "std::vector" is used and CUDA kernels are involved in the main codes, it is strongly recommended to add test cases for multiple settings, especially the corner cases.Since "std::vector" is used and CUDA kernels are involved in the main codes, it is strongly recommended to add test cases for multiple settings, especially the corner cases.